server/web_ui/src/components/totpdisplay.rs

```diff
@@ -0,0 +1,343 @@
+// Alert occurs at 5s
+static ALERT_THRESHOLD: f32 = 0.25;
+
+enum TotpStatus {
+    Waiting,
+    Secret(u32),
+}
+
+pub struct TotpDisplayApp {
+    secret: TotpStatus,
+    step: u32,
+    main_timer: Option<Timeout>,
+    ticks_timer: Option<Interval>,
+    sync_timer: Option<Timeout>,
+    ticks_left: u8,
+}
+
+pub enum Msg {
+    FetchTotpAndResetTimer,
+    NewTotp(u32),
+    Tick,
+    StartTicking,
+    TotpConfirmed,
+    TotpNotConfirmed,
+    Cancel,
+    InvalidState,
+}
+
+#[derive(Properties, PartialEq)]
+pub struct TotpProps {
+    pub state: IdentifyUserState,
+    pub other_id: String,
+    pub cb: Callback<IdentifyUserTransition>,
+}
+
+impl TotpDisplayApp {
+    async fn renew_totp(other_id: String) -> Msg {
+        let uri = format!("/v1/person/{}/_identify_user", other_id);
+        let request = IdentifyUserRequest::DisplayCode;
+        let Ok(state_as_jsvalue) = serde_json::to_string(&request)
+            .map(|s| JsValue::from(&s))
+             else {
+            return Msg::Cancel
+        };
+        let response = match do_request(&uri, RequestMethod::POST, Some(state_as_jsvalue)).await {
+            Ok((_, _, response, _)) => response,
+            Err(_) => return Msg::Cancel,
+        };
+        match serde_wasm_bindgen::from_value(response) {
+            // TODO: check how the backend encodes the totp when sending it
+            Ok(IdentifyUserResponse::ProvideCode { totp, step: _ }) => Msg::NewTotp(totp),
+            _ => Msg::Cancel,
+        }
+    }
+
+    fn get_time_left_from_now(&self) -> u32 {
+        #[allow(clippy::expect_used)]
+        let dur = SystemTime::now()
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .expect("invalid duration from epoch now");
+        let secs: u128 = dur.as_millis();
+        let step = self.step as u128;
+        (step * 1000 - secs % (step * 1000)) as u32
+    }
+
+    fn get_time_left_from_now_selfless(step: u128) -> u32 {
+        #[allow(clippy::expect_used)]
+        let dur = SystemTime::now()
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .expect("invalid duration from epoch now");
+        let secs: u128 = dur.as_millis();
+        (step * 1000 - secs % (step * 1000)) as u32
+    }
+
+    fn get_ring_color(&self, time_remaining: u32) -> AttrValue {
+        // it's a bit hacky but we want it to be green starting from 0 (aka no totp ring) so by the next second
+        // (aka when it goes to 30) it's already 100% green, since the transition takes 1s we have to start doing this from
+        // time_remaining == 1
+        AttrValue::from(if time_remaining <= 1 {
+            "green"
+        } else if time_remaining <= (ALERT_THRESHOLD * self.step as f32) as u32 {
+            "red"
+        } else if time_remaining <= (WARNING_THRESHOLD * self.step as f32) as u32 {
+            "orange"
+        } else {
+            "green"
+        })
+    }
+
+    // TODO! It's not remotely doing what I wanted to do, damned CSS
```

Comment:
Does it work or not? 🤔

Reply:
Errrr let's say it's not harming anyone for the time being. I mean it's a very hacky solution to get the transition to work the way I'd want it, but it's very unreliable and most of the times it's as if it didn't exist. If we remove it everything else should just work fine